src/Membership/MembershipRequestHandler.php

```diff
@@ -48,38 +49,61 @@ public function __construct(
         $this->committeeManager = $committeeManager;
     }
 
-    public function handle(MembershipRequest $membershipRequest)
+    public function handle(MembershipRequest $membershipRequest): Adherent
     {
         $adherent = $this->adherentFactory->createFromMembershipRequest($membershipRequest);
+        $this->manager->persist($adherent);
+        $this->sendEmailValidation($adherent);
+
+        $this->dispatcher->dispatch(UserEvents::USER_CREATED, new UserEvent($adherent));
+
+        return $adherent;
+    }
+
+    public function sendEmailValidation(Adherent $adherent): void
+    {
         $token = AdherentActivationToken::generate($adherent);
 
-        $this->manager->persist($adherent);
         $this->manager->persist($token);
         $this->manager->flush();
 
         $activationUrl = $this->generateMembershipActivationUrl($adherent, $token);
         $this->mailer->sendMessage(AdherentAccountActivationMessage::createFromAdherent($adherent, $activationUrl));
+    }
+
+    public function register(MembershipRequest $membershipRequest): void
+    {
+        $adherent = $this->adherentFactory->createFromMembershipRequest($membershipRequest);
+        $this->manager->persist($adherent);
+        $adherent->join();
+        $this->sendEmailValidation($adherent);
 
         $this->dispatcher->dispatch(UserEvents::USER_CREATED, new UserEvent($adherent));
+        $this->dispatcher->dispatch(AdherentEvents::REGISTRATION_COMPLETED, new AdherentAccountWasCreatedEvent($adherent, $membershipRequest));
```

Comment:
We always have 2 emails sent when you register as an Adherent if I am not mistaken ? We should have only 1 no ? Check what is listening to this event.

Reply:
In the registration.feature it is tested that we have only 1 mail send. don t think if the test pass we have an other hidden email :s

Reply:
It does test if a particular email is sent but it does not sent if any other email is sent.
So you could add a check to test if the other email is not sent.

Reply:
it test the number of email sent

Reply:
@cacahouwete: no. it tests that a message X is indeed sheduled. it does not assert that a message Y has been scheduled or not.

Reply:
I have create behat rule to check nb emails in DB